tools/wake-format/formatter.h

```diff
@@ -306,15 +320,20 @@ struct Formatter {
         action, WalkPredicateAction<Walker, Predicate>(texas_ranger, predicate))};
   }
 
+  template <class FMT>
+  Formatter<SeqAction<Action, WalkChildrenAction<FMT>>> walk_children(FMT formatter) {
+    return {SeqAction<Action, WalkChildrenAction<FMT>>(action, WalkChildrenAction<FMT>{formatter})};
+  }
+
   template <class F>
   Formatter<SeqAction<Action, EscapeAction<F>>> escape(F f) {
     return {SeqAction<Action, EscapeAction<F>>(action, EscapeAction<F>{f})};
   }
 
-  wcl::doc format(ctx_t ctx, CSTElement node) {
+  wcl::doc format(ctx_t ctx, CSTElement node, bool assert_empty = true) {
```

Comment:
The main use case is inside of a "sub formatter" where we don't actually fully consume the node but I think it would be better to just not have it. I'm going to see if there is a way I can make these cases fully consume the node. I think the code before wasn't

Reply:
okay, so that was much harder to fix than expected and became a tiny bit of a yak shave but its resolved and uploaded now.

There is still one case where it is needed for the `walk_children` since its actually node expected that `child` is empty until all children have been processed

Reply:
So `format` seems like it should do the check but `compose` should not. Since a formatter passed to something like `fmt_if` probably won't consume all of its input but that just gets composed into the outerone where format is called where it *should* do that....IDK does that sound right?

Reply:
chatted about this in our 1:1, has been removed